### PR TITLE
Bug 1324557 - Collect stats for sync engines for telemetry (excluding bookmarks)

### PR DESCRIPTION
**Pull request by sleroux**
First pass at adding a way of collection telemetry information from our sync engines. From a high level,

A `SyncStatsReport` contains all of the telemetry information related to a single sync operation. Within the report, there are many `SyncEngineStats` which contain information about each of our sync engines. 

A report is started when a sync operation starts. For each engine, a `SyncEngineStatsObserver` is passed to each engine which acts as the engine's `SyncStatsDelegate`. During the course of the engine's operations, the relevant data is passed to the delegate through lifecycle events (willBegin/didUpload/didDownload/DidFinished). As each engine completes, the `SyncEngineStats` appended to the in-progress report.

When the syncing operation has completed, the report is finalized and broadcasted using an `NSNotification` and closed.

This patch contains adding the abstractions for storing sync information, report creation, and collecting stats for each engine except for bookmarks. Adding the stats delegate into the merger seems a bit hairy so before I move forward I'd like to know if @rnewman has any thoughts on the WIP patch as well as threading in the stats collection into bookmarks.

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 18 | Colons should be next to the identifier when specifying a type. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 23 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by sleroux** [comment]
Also to note:

1. Validation errors from bookmarks will be in another patch for https://bugzilla.mozilla.org/show_bug.cgi?id=1324558.
2. Actually sending the telemetry will probably happen else where and will rely on serializing the `NSNotification` report from the sync client to send to the servers.

File: Providers/Profile.swift [hunk 1]
```diff
@@ -703,15 +703,22 @@ public class BrowserProfile: Profile {
 
         // Used as a task queue for syncing.
         private var syncReducer: AsyncReducer<EngineResults, EngineTasks>?
+        private var syncStatsReport: SyncStatsReport?
```
Comment (rnewman) [inline Providers/Profile.swift]: I'm not a huge fan of this shared mutable conditional state. Can we not thread this into a sync and back out the end as part of a result?

File: Providers/Profile.swift [hunk 2]
```diff
@@ -703,15 +703,22 @@ public class BrowserProfile: Profile {
...
 
         private func beginSyncing() {
             notifySyncing(NotificationProfileDidStartSyncing)
+            if let account = profile.account {
+                syncStatsReport = SyncStatsReport(when: NSDate.now(), account: account, didLogin: false, why: "")
```
Comment (rnewman) [inline Providers/Profile.swift]: Perhaps use a server timestamp?

File: Providers/Profile.swift [hunk 3]
```diff
@@ -794,6 +806,11 @@ public class BrowserProfile: Profile {
             NSNotificationCenter.defaultCenter().postNotification(NSNotification(name: notification, object: syncDisplayState?.asObject()))
         }
 
+        private func notifySyncStatsReport(report: SyncStatsReport) {
```
Comment (rnewman) [inline Providers/Profile.swift]: `finishAndNotify…`

File: Providers/Profile.swift [hunk 4]
```diff
@@ -1082,33 +1099,54 @@ public class BrowserProfile: Profile {
         }
 
         private func syncClientsWithDelegate(delegate: SyncDelegate, prefs: Prefs, ready: Ready) -> SyncResult {
-            log.debug("Syncing clients to storage.")
-            let clientSynchronizer = ready.synchronizer(ClientsSynchronizer.self, delegate: delegate, prefs: prefs)
-            return clientSynchronizer.synchronizeLocalClients(self.profile.remoteClientsAndTabs, withServer: ready.client, info: ready.info)
+            return statsWrapEngineSync("clients") { statsDelegate in
```
Comment (rnewman) [inline Providers/Profile.swift]: Why not have the `SyncDelegate` also be the stats delegate, and have the stats API encapsulate the engine part? So these functions would all stay the same, then elsewhere, at point of use, you'd "check out" a stats recorder for the implementing engine from the delegate.

File: Sync/SyncTelemetry.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+public protocol SyncStatsDelegate: class {
+    func engineWillBeginCollectingStats()
+    func engineDidGenerateUploadStats(stats: SyncUploadStats)
+    func engineDidGenerateApplyStats(stats: SyncDownloadStats)
+    func engineDidFinishCollectingStats(status: SyncStatus) -> SyncEngineStats?
+}
+
+// Delegate object that is passed along to each synchronizer to pull out upload/downloading stats
+public class SyncEngineStatsObserver: SyncStatsDelegate {
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: I think you should fetch this from the delegate at point of use.

File: Sync/Synchronizers/HistorySynchronizer.swift [hunk 6]
```diff
@@ -93,7 +94,15 @@ public class HistorySynchronizer: IndependentRecordSynchronizer, Synchronizer {
             // We apply deletions immediately. Yes, this will throw away local visits
             // that haven't yet been synced. That's how Sync works, alas.
             if payload.deleted {
-                return storage.deleteByGUID(guid, deletedAt: modified).bind(maskSomeFailures)
+                stats.applied += 1
+                return storage.deleteByGUID(guid, deletedAt: modified).bind({ result in
```
Comment (rnewman) [inline Sync/Synchronizers/HistorySynchronizer.swift]: Use `effect` here to handle side-effects and pass through the result.

File: Sync/Synchronizers/LoginsSynchronizer.swift [hunk 7]
```diff
@@ -89,37 +89,49 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
...
                 log.warning("Login record \(guid) is invalid. Skipping.")
+                stats.failed += 1
                 return succeed()
             }
 
             // We apply deletions immediately. That might not be exactly what we want -- perhaps you changed
             // a password locally after deleting it remotely -- but it's expedient.
             if payload.deleted {
-                return storage.deleteByGUID(guid, deletedAt: rec.modified)
+                return storage.deleteByGUID(guid, deletedAt: rec.modified).bind(recordApplyResult)
```
Comment (rnewman) [inline Sync/Synchronizers/LoginsSynchronizer.swift]: `effect` again. Simplifies `recordApplyResult`, too.

**Comment by rnewman** [comment]
For bookmarks, note that failures can occur at several points:

- An incoming record can be completely invalid, and won't even reach the buffer. (You might have ignored this for other engines.)
- Incoming records can be buffered, but the buffer is known to be inconsistent, so we neither attempt to apply the buffer nor upload any records.
- The buffer can appear to be valid, but further issues are found while producing the merge tree or walking the merge tree. Again, we abort immediately.

The resulting merge tree is then walked to produce operations, and those operations are run against the server, mirror, local, and buffer. Any one of those operations can fail due to the records/data, a bug, or external circumstances.

The question then becomes: what are you trying to measure?

This is my outstanding criticism of the counts that desktop has always logged: they don't really tell you anything. Okay, so we downloaded ten records and failed to apply five. So what?

We can identify a collection of different counts for bookmarks:

- We think the record we downloaded was malformed.
- Something went wrong when handling a downloaded record, prior to buffering.
- Something went wrong when buffering this batch of N records.
- We produced an op with 5 things to copy from buffer to mirror, 3 things that were merged in memory, etc. etc.
- The server rejected 7 of the 15 records we uploaded.
- We failed to write N changes to one of our operation destinations, aborting the whole sync.

But there is a bunch of information hiding inside stuff we can't count:

- The buffer is inconsistent. Note that this is an emergent property of a collection of records — you can't say we downloaded 23 records and failed to apply 23.
- The buffer is easy-consistent but we couldn't build a tree out of it. Again, this is an emergent property, and we short-circuit to avoid burning CPU on the same checks every time.
- An op failed for some reason.

What are we trying to measure? Rate of change of bookmarks? Number of retries? Something else?

File: Providers/Profile.swift (hunk 1 above)
Comment (sleroux) [inline Providers/Profile.swift]: I'm not a fan of it either. I'll see if I can weave it in to fit in with the more functional code in this module.

File: Providers/Profile.swift (hunk 2 above)
Comment (sleroux) [inline Providers/Profile.swift]: That would be ideal. How would you suggest getting one? Is there a 'start' ping of some sort that can give me a timestamp?

Comment (rnewman) [inline Providers/Profile.swift]: The first thing that happens in a sync is we fetch `info/collections` to determine what's changed on the server. That fetch has the server timestamp in a header, and makes for an excellent symbolic beginning to a sync.

File: Providers/Profile.swift (hunk 4 above)
Comment (sleroux) [inline Providers/Profile.swift]: I wasn't sure if I wanted to conflate the two since it wasn't clear what SyncDelegate was around for. Moving the stats stuff into it will clean this PR up quite a bit.

**Comment by sleroux** [comment]
> This feels pretty messy, so I'd like to see time spent making it cleaner/clearer/less likely to be wrong.

One thing I noticed in the desktop implementation is that their collection is only in one central place which handles everything about a collection [1] where-as we have the logic implemented in unique ways across our different engines. I don't think that's necessarily a bad thing since each engine is somewhat unique but it would be nice if we can reduce the amount of similar code across all of these engines to reduce errors. Any thoughts about consolidating this across engines?

[1] https://dxr.mozilla.org/mozilla-central/source/services/sync/modules/engines.js#748

**Comment by sleroux** [comment]
Comments for re-flagging feedback:

I went back to the drawing board with this patch and cleaned up how we were collecting sync stats. Instead of passing through a SyncStatsDelegate to all of the engines, each engine creates a `SyncEngineStatsSession` [1] which captures the download/upload stats along with timing the sync. When the sync is finished, the session object is passed alongside the .Completed SyncStatus enum as an associated property [2]. Once the engine finishes, all of the results are bubbled through to the existing `endSyncing` call where the stats are unpacked and logged (for now). Instead of having the `SyncOperationStatsSession` be a shared mutable property, it gets created and passed through the deferreds from `syncSeveral`/`syncWith`. I've also exposed the `InfoCollections` metadata so I can grab the returned server timestamp to use for stats session. Lastly, I've threaded in a `why` enum which tells the session why a sync was started.

[1] https://github.com/mozilla-mobile/firefox-ios/pull/2324/files#diff-a7d3d9920fa098baa4c1802c71828d4aR158
[2] https://github.com/mozilla-mobile/firefox-ios/pull/2324/files#diff-a7d3d9920fa098baa4c1802c71828d4aR79
[3] https://github.com/mozilla-mobile/firefox-ios/pull/2324/files#diff-7a45ce0d70d55bdf297fcacd47e4e874R1211

File: ClientTests/MockProfile.swift [hunk 8]
```diff
@@ -21,11 +21,15 @@ public class MockSyncManager: SyncManager {
...
-    public func syncClients() -> SyncResult { return deferMaybe(.Completed) }
-    public func syncClientsThenTabs() -> SyncResult { return deferMaybe(.Completed) }
-    public func syncHistory() -> SyncResult { return deferMaybe(.Completed) }
-    public func syncLogins() -> SyncResult { return deferMaybe(.Completed) }
-    public func syncEverything() -> Success {
+    private func completedWithStats(collection: String) -> Deferred<Maybe<SyncStatus>> {
+        return deferMaybe(SyncStatus.Completed(SyncEngineStatsSession(collection: collection)))
+    }
+    
+    public func syncClients() -> SyncResult { return completedWithStats("clients") }
```
Comment (rnewman) [inline ClientTests/MockProfile.swift]: I think for these mocks you should have "mock" in the string.

File: Providers/Profile.swift [hunk 9]
```diff
@@ -722,24 +723,41 @@ public class BrowserProfile: Profile {
             notifySyncing(NotificationProfileDidStartSyncing)
         }
 
-        private func endSyncing(result: Maybe<EngineResults>?) {
+        private func endSyncing(result: SyncOperationResult) {
             // loop through status's and fill sync state
```
Comment (rnewman) [inline Providers/Profile.swift]: s/status's/statuses

File: Providers/Profile.swift [hunk 10]
```diff
@@ -722,24 +723,41 @@ public class BrowserProfile: Profile {
...
 
-            reportEndSyncingStatus(syncDisplayState, engineResults: result)
+            reportSyncPingForResult(result)
             notifySyncing(NotificationProfileDidFinishSyncing)
             syncReducer = nil
         }
 
-        private func reportEndSyncingStatus(displayState: SyncDisplayState?, engineResults: Maybe<EngineResults>?) {
+        private func reportSyncPingForResult(opResult: SyncOperationResult) {
+            guard profile.prefs.boolForKey("settings.sendUsageData") ?? true else {
```
Comment (rnewman) [inline Providers/Profile.swift]: If this code exists in more than one place in the code, lift it into a method on `profile`?

File: Providers/Profile.swift [hunk 11]
```diff
@@ -1208,7 +1229,8 @@ public class BrowserProfile: Profile {
         }
 
         // This SHOULD NOT be called directly: use syncSeveral instead.
-        private func syncWith(synchronizers: [(EngineIdentifier, SyncFunction)]) -> Deferred<Maybe<[(EngineIdentifier, SyncStatus)]>> {
+        private func syncWith(synchronizers: [(EngineIdentifier, SyncFunction)],
+                              statsSession: SyncOperationStatsSession) -> Deferred<Maybe<[(EngineIdentifier, SyncStatus)]>> {
             guard let account = self.profile.account else {
```
Comment (rnewman) [inline Providers/Profile.swift]: I suspect that we should be capturing `self.profile.account` in `syncSeveral`, doing this check there, and passing it as an argument to `syncWith`, alongside the stats session. That allows us to construct the `statsSession` from the account directly in `syncSeveral`, which will make it less likely that someone will forget to set its fields…

File: Providers/Profile.swift [hunk 12]
```diff
@@ -1217,6 +1239,9 @@ public class BrowserProfile: Profile {
                 return deferMaybe(statuses)
             }
 
+            statsSession.uid = account.uid
+            statsSession.deviceID = account.deviceRegistration?.id
```
Comment (rnewman) [inline Providers/Profile.swift]: … these.

File: Sync/SyncStateMachine.swift [hunk 13]
```diff
@@ -220,10 +220,8 @@ public class BaseSyncState: SyncState {
     var scratchpad: Scratchpad
 
     // TODO: 304 for i/c.
-    public func getInfoCollections() -> Deferred<Maybe<InfoCollections>> {
-        return chain(self.client.getInfoCollections(), f: {
-            return $0.value
-        })
+    public func getInfoCollectionResponse() -> Deferred<Maybe<StorageResponse<InfoCollections>>> {
```
Comment (rnewman) [inline Sync/SyncStateMachine.swift]: `getInfoCollectionsResponse`

File: Sync/SyncTelemetry.swift [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+import Foundation
+import Shared
+import Account
+
+public enum SyncReason: String {
+    case Startup = "startup"
+    case Scheduled = "scheduled"
+    case Backgrounded = "backgrounded"
+    case User = "user"
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Call this `SyncNow`? There's a separate code path for pull-to-refresh in the History panel.

File: Sync/Synchronizers/HistorySynchronizer.swift (hunk 6 above)
Comment (rnewman) [inline Sync/Synchronizers/HistorySynchronizer.swift]: Still should use `effect`.

**Comment by sleroux** [comment]
Weird. I can't comment on the last comment you made in the review. Anyways,

> Still should use effect.

I don't think I can use `effect` here because I need the result value to check to see if it failed. If the deletion failed then the effect block won't get invoked.

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Errors

File | Line | Reason |
| --- | ----- | ----- |
Mocking.swift | 31 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Providers/Profile.swift [hunk 15]
```diff
@@ -677,34 +678,38 @@ open class BrowserProfile: Profile {
...
-            // loop through status's and fill sync state
+        fileprivate func endSyncing(_ result: SyncOperationResult) {
+            // loop through statuses and fill sync state
             syncLock.lock()
             defer { syncLock.unlock() }
             log.info("Ending all queued syncs.")
 
-            if let syncResult = result {
-                syncDisplayState = SyncStatusResolver(engineResults: syncResult).resolveResults()
+            syncDisplayState = SyncStatusResolver(engineResults: result.engineResults).resolveResults()
```
Comment (rnewman) [inline Providers/Profile.swift]: There might be a subtle little bug here.

If we previously called `endSyncing` with `None`, we wouldn't change the display state.

Now we pass the none-`Maybe` to `SyncStatusResolver`, which resolves to `SyncDisplayState.Bad(message: nil)`.

Is this change intentional?

File: Providers/Profile.swift [hunk 16]
```diff
@@ -1178,29 +1202,32 @@ open class BrowserProfile: Profile {
             }
             
             return readyDeferred >>== self.takeActionsOnEngineStateChanges >>== { ready in
-                function(delegate, self.prefsForSync, ready)
+                // Once we are ready and have a server timestamp to begin our session, start recording
```
Comment (rnewman) [inline Providers/Profile.swift]: Nit: period.

File: Sync/SyncTelemetry.swift [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+// TODO(sleroux): Implement various bookmark validation issues we can run into
+public struct ValidationStats: Stats {
+    public func hasData() -> Bool {
+        return false
+    }
+}
+
+public class StatsSession {
+    private var took: UInt64 = 0
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Make this an `Int64`. We don't expect a sync to take more than 9,223,372,036,854,775 seconds, and it protects us from clock errors.

File: Sync/SyncTelemetry.swift [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    private var startTime: Timestamp?
+
+    public func start(time: Timestamp = Date.now()) {
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Unless you specifically need to report wall clock instants, use uptime here. That way you're guaranteed a monotonically increasing clock, and you won't find syncs taking a negative amount of time just because the device clock shifted.

E.g.,

https://developer.apple.com/reference/foundation/processinfo/1414553-systemuptime

or

https://developer.apple.com/reference/dispatch/dispatchtime/2300047-uptimenanoseconds

File: Sync/SyncTelemetry.swift [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+        return startTime != nil
+    }
+
+    public func end() -> Self {
+        guard let startTime = startTime else {
+            assertionFailure("SyncOperationStats called end without first calling start!")
+            return self
+        }
+
+        took = Date.now() - startTime
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Note that at this point if the wall clock has jumped backwards, Firefox will crash. Be very, very careful when using `UInt64`!

File: Sync/SyncTelemetry.swift [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+        return self
+    }
+}
+
+// Stats about a single engine's sync
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Nit: closing periods.

File: Sync/Synchronizers/IndependentRecordSynchronizer.swift [hunk 21]
```diff
@@ -37,6 +37,23 @@ class Uploader {
 }
 
 open class IndependentRecordSynchronizer: TimestampedSingleCollectionSynchronizer {
+    private func reportApplyStatsWrap<T>(apply: @escaping (T) -> Success) -> (T) -> Success {
+        return { record in
+            var stats = SyncDownloadStats()
+            stats.applied = 1
```
Comment (rnewman) [inline Sync/Synchronizers/IndependentRecordSynchronizer.swift]: Move these two lines inside the `bind`, for casual thread safety concerns.

File: Sync/SyncTelemetry.swift (hunk 18 above)
Comment (sleroux) [inline Sync/SyncTelemetry.swift]: Since we want milliseconds and `ProcessInfo` returns seconds I opted for `uptimeNanoseconds`. I also found this handy function on UInt64 to help with the negative/overflow issue:

https://developer.apple.com/reference/swift/uint64/1541211-subtractwithoverflow

I'm able to detect the clock jumping backwards to prevent any crashes and log it to the logger.

Comment (sleroux) [inline Sync/SyncTelemetry.swift]: According to the sync ping spec [1] the ping expects this to be milliseconds from epoch instead of uptime since boot which is what `uptimeNanoseconds` returns. I've pushed this using uptime but I think I should be using .now() instead...

[1] http://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/data/sync-ping.html

Comment (rnewman) [inline Sync/SyncTelemetry.swift]: `uptimeNanoseconds` should be monotonically increasing, so you should never have to handle overflow. But using `Int64` is a worthwhile thing to do anyway.

Comment (rnewman) [inline Sync/SyncTelemetry.swift]: "Unless you specifically need to report wall clock instants" applies, then.

You should use wall clock to report start time, and you should use a _different_ monotonically increasing clock to calculate duration. You should not use the wall clock to calculate duration.

**Comment by sleroux** [comment]
> I'm not a huge fan of having to thread this kind of state through a program, but waddaya gonna do?

I'm not a fan of it either TBH. I had a hard time coming up with alternatives though. Any suggestions? I think the patch is fine for landing but just for curiosity.

File: Providers/Profile.swift (hunk 15 above)
Comment (sleroux) [inline Providers/Profile.swift]: Oddly enough I don't think the previous logic for the `None` case is ever used. `syncReducer`s terminal returns a `Deferred<Maybe<()>>` so the old `endSyncing` never got called with an optional.

**Comment by rnewman** [comment]
> I'm not a fan of it either TBH. I had a hard time coming up with alternatives though. Any suggestions? I think the patch is fine for landing but just for curiosity.

Four possibilities.

1. This is the classic use for aspects. Rather than add your own timing code, you implement some kind of wrapper/interceptor that calls through to the original code and times it. Unfortunately, you don't just want timing and success, you want detailed operation counts (behavioral logging), and that's difficult.

2. If you can phrase work done in terms of states, then you can isolate the telemetry. E.g., if record application weren't done record-by-record, but instead work was accumulated to be applied in one shot (similar to the bookmark merge ops), then you can get your succeeded/failed by looking at the inputs and outputs of the process.

3. Similarly, you could make each 'sink' (e.g., the history store) a kind of stateful source of telemetry data. Have it count the operations performed on it. Now you can get your telemetry data at the end of a sync just by poking storage. This assumes that storage operations are the right granularity: your storage layer now needs to know the kinds of telemetry you gather, rather than the sync engine knowing it. Imagine if your telemetry was "added database rows" — SQLite gives you that already!

4. Rather than threading an accumulator through your application code, decouple via notifications/broadcasts or another kind of information 'leakage' channel. For example, you could have a tri-state flag (not touched, failed, succeeded) on each input record, or you could send a notification each time a record was processed. Again, this allows you to do your telemetry work outside the main code. However, you'll note that you need to introduce some kind of scoped/global state — the message channel — and implement an ad hoc message dispatch system that dynamically couples the processor and the consumer, so this isn't free.

TANSTAAFL.

**Comment by sleroux** [comment]
Thanks for the indulgence.

> This is the classic use for aspects. Rather than add your own timing code, you implement some kind of wrapper/interceptor that calls through to the original code and times it. Unfortunately, you don't just want timing and success, you want detailed operation counts (behavioral logging), and that's difficult.

I've seen this done in other iOS applications using method swizzling but I agree - keeping track of the sessions would be difficult. I'm also not a fan of swizzlin' as it usually leads to sizzling bugs.

I had  WIP patch that used NSNotifications to implement (4) but it felt too disjoint from the sync code since it wasn't as clear when a sync as a whole started and ended.

I like 3/4 though. From what I understand the goal is to move the stateful-ness of telemetry into the stateful layer of the code and make inferences from the app level. To your point of granularity it would seem that this would also be the most accurate approach as well.
